torch/_inductor/comm_analysis.py

```diff
@@ -0,0 +1,179 @@
+import math
+
+from .utils import get_dtype_size
+from . import ir
+from .utils import sympy_product
+from .virtualized import V
+
+ncclFuncAllReduce = "allreduce"
+ncclFuncAllGather = "allgather"
+ncclFuncReduceScatter = "reducescatter"
+
+NCCL_HW_NVLINK = 0
+NCCL_HW_PCI = 1
+NCCL_HW_NET = 2
+
+NCCL_ALGO_TREE = 0
+NCCL_ALGO_RING = 1
+
+# Latencies in us
+baseLat = [
+    6.8,  # Tree (LL)
+    6.6,  # Ring (LL)
+]
+# Latencies in us
+hwLat = [
+    # NVLINK
+    [
+        .6,  # Tree (LL)
+        .6,  # Ring (LL)
+    ],
+    # PCI
+    [
+        1.0,  # Tree (LL)
+        1.0,  # Ring (LL)
+    ],
+    # NET
+    [
+        5.0,  # Tree (LL)
+        2.7,  # Ring (LL)
+    ],
+]
+
+VOLTA_COMPCAP_IDX = 0
+AMPERE_COMPCAP_IDX = 1
+HOPPER_COMPCAP_IDX = 2
+
+# LL128 max BW per channel
+llMaxBws = [
+    # Volta-N1/Intel-N2/Intel-N4
+    [39.0, 39.0, 20.4],
+    # Ampere-N1/AMD-N2/AMD-N4
+    [
+        87.7,
+        22.5,  # avg of ring & tree
+        19.0,
+    ],
+    # Hopper-N1/AMD-N2/AMD-N4
+    [
+        87.7,
+        22.5,  # avg of ring & tree
+        19.0,
+    ],
+]
+
+bwNVLINK = 300  # unit: GB/s, uni-directional P2P bandwidth per card
+bwInfiniBand = 25  # unit: GB/s, uni-directional P2P bandwidth per node
+
+
+def get_collective_type(snode: "scheduler.BaseSchedulerNode") -> str:
+    if isinstance(snode.node, (ir.AllReduce, ir.AllReduceCoalesced)):
+        return ncclFuncAllReduce
+    elif isinstance(snode.node, (ir.AllGatherIntoTensor, ir.AllGatherIntoTensorCoalesced)):
+        return ncclFuncAllGather
+    elif isinstance(snode.node, (ir.ReduceScatterTensor, ir.ReduceScatterTensorCoalesced)):
+        return ncclFuncReduceScatter
+    else:
+        raise Exception(f"Unsupported collective type: {snode.node}")
+
+
+def estimate_nccl_collective_runtime(snode: "scheduler.BaseSchedulerNode") -> float:
+    """
+    Returns estimated NCCL collective runtime in nanoseconds (ns).
+
+    The following heuristics are copied from https://github.com/NVIDIA/nccl/blob/master/src/graph/tuning.cc.
+    We aim to estimate the runtime as accurately as possible.
+
+    Assumptions:
+    - only ring algorithm (NCCL_ALGO_RING) is used
+    - only Low-Latency protocol (NCCL_PROTO_LL) is used, i.e. Simple or LL128 is not used
+    - only A100 gpu
```

Comment:
This looks great! 

As a follow up we might can see how to get those numbers from some NCCL APIs directly instead of hard code them, as it might be fixed for one hardware

Reply:
agreed - the original NCCL code has V100 and H100 numbers too, we can expand this in follow-up, or find a way to have an NCCL API for it